bin/render-activity-stream-html.js

```diff
@@ -85,33 +71,48 @@ function templateHTML(options, html) {
   if (isPrerendered) {
     scripts.unshift(`${options.baseUrl}prerendered/static/activity-stream-initial-state.js`);
   }
-  const scriptTag = `
-    <script>
-// Don't directly load the following scripts as part of html to let the page
-// finish loading to render the content sooner.
-for (const src of ${JSON.stringify(scripts, null, 2)}) {
-  // These dynamically inserted scripts by default are async, but we need them
-  // to load in the desired order (i.e., bundle last).
-  const script = document.body.appendChild(document.createElement("script"));
-  script.async = false;
-  script.src = src;
+
+  return scripts;
 }
-    </script>`;
+
+/**
+ * templateHTML - Generates HTML for activity stream, given some options and
+ * prerendered HTML if necessary.
+ *
+ * @param  {obj} options
+ *         {str} options.locale         The locale to render in lang="" attribute
+ *         {str} options.direction      The language direction to render in dir="" attribute
+ *         {str} options.baseUrl        The base URL for all local assets
+ *         {bool} options.debug         Should we use dev versions of JS libraries?
+ *         {bool} options.noscripts     Should we include scripts in the prerendered files?
+ * @param  {str} html    The prerendered HTML created with React.renderToString (optional)
+ * @return {str}         An HTML document as a string
+ */
+function templateHTML(options, html) {
+  const isPrerendered = !!html;
+  const loadScriptTag = `<script src="${options.baseUrl}data/content/activity-stream-load-scripts.js"></script>`;
+  const renderInfo = [
+    isPrerendered && "prerendered",
+    options.noscripts ? "noscripts" : "scripts-loaded",
+    options.debug && "debug",
+  ].filter(i => i);
   return `<!doctype html>
 <html lang="${options.locale}" dir="${options.direction}">
+  <!-- Render info: ${renderInfo.join(", ")} -->
   <head>
     <meta charset="utf-8">
-    <meta http-equiv="Content-Security-Policy" content="default-src 'none'; script-src 'unsafe-inline' resource: chrome:; connect-src https:; img-src https: data: blob:; style-src 'unsafe-inline';">
+    <meta http-equiv="Content-Security-Policy" content="default-src 'none'; object-src 'none'; script-src resource: chrome:; connect-src https:; img-src https: data: blob:; style-src 'unsafe-inline';">
     <title>${options.strings.newtab_page_title}</title>
+    <meta name="scripts" content="${getScripts(options, isPrerendered).join(",")}" />
```

Comment:
Interesting approach… seems a little scary having a script that loads more, but I guess it works. Test is complaining because it's only looking for quoted chrome/resource uris. I think if you stringify, you'll happen to trick the test to be happy. I.e., `content='["resource://activity-stream/…","resource:…"]'` instead of `content="resource://activity-stream/…,resource:…"`

https://searchfox.org/mozilla-central/rev/d302c3058330a57f238be4062fddea629311ce66/browser/base/content/test/static/browser_all_files_referenced.js#421-424

Reply:
Ack. The logic there only checks for resource uris only if there are no chrome uris found on the line. I suppose you can just render as html comments the scripts that are included in the meta ?

Reply:
This makes the test happy…
```js
    <!-- Dynamically load scripts:
${getScripts(options, isPrerendered).map(s => `"${s}"`).join("\n")}
    -->
```